tsdb/head.go

```diff
@@ -624,24 +623,6 @@ func (h *Head) Init(minValidTime int64) (err error) {
 		}
 	}()
 
-	// If sharding is enabled, disable it while initializing, and calculate the shards later.
-	// We're going to use that field for other purposes during WAL replay,
-	// so we don't want to waste time on calculating the shard that we're going to lose anyway.
-	if h.opts.EnableSharding {
-		h.opts.EnableSharding = false
-		defer func() {
-			h.opts.EnableSharding = true
-			if err == nil {
```

Comment:
Removing this condition should suffice to fix the bug.

Reply:
I think the design is bad. We shouldn't use the same field for two different unrelated purposes just to save memory. I don't want to reason about all possible edge cases caused by doing it. I propose to revert not just the fix the bug, but also to fix the desing.